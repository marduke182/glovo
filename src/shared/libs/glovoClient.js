import Axios from 'axios';
import { addErrorCatching } from './axiosInterceptors';

const glovoClient = Axios.create({
  baseURL: `/api/`,
});

export default addErrorCatching(glovoClient);

