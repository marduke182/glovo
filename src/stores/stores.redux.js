import { createSelector } from 'reselect';
import { createAction, handleActions } from 'redux-actions';

import glovoClient from '@/shared/libs/glovoClient';

import mapStore from './libs/mapStore';
import sortByOpened from '@/stores/libs/sortByOpened';

const fetchStoreRequest = createAction('FETCH_STORE_REQUEST');
const fetchStoreSuccess = createAction('FETCH_STORE_SUCCESS');
const fetchStoreFailure = createAction('FETCH_STORE_FAILURE');

const changeTag = createAction('CHANGE_TAG');

const initialState = {
  loading: 0, // 0 means no loading
  error: null,
  tag: '',
  storeByCategory: {}
};

export default handleActions(
  {
    [fetchStoreRequest]: state => ({
      ...state,
      loading: state.loading + 1
    }),
    [fetchStoreFailure]: (state, { payload: { error } }) => ({
      ...state,
      loading: Math.max(state.loading - 1, 0),
      error
    }),
    [fetchStoreSuccess]: (state, { payload: { category, stores } }) => ({
      ...state,
      loading: Math.max(state.loading - 1, 0),
      storeByCategory: {
        ...state.storeByCategory,
        [category]: stores
      },
      error: null
    }),
    [changeTag]: (state, { payload: tag}) => ({
      ...state,
      tag
    })
  },
  initialState
);

export const name = 'stores';

// ACTIONS
const getStoresOf = category => async dispatch => {
  dispatch(fetchStoreRequest());

  const { data, error } = await glovoClient.get(`stores?category=${category}`);

  if (error) {
    dispatch(fetchStoreFailure({ error }));
    return;
  }

  dispatch(
    fetchStoreSuccess({
      category,
      stores: data.stores
    })
  );
};

// Basic Selectors SELECTORS
const loadingSelector = state => state[name].loading !== 0;
const errorSelector = state => state[name].error;
const tagSelector = state => state[name].tag;
const storesByCategory = (state, props) => state[name].storeByCategory[props.category] || [];

// Get all tags from the stores in a category
function onlyUnique(value, index, self) {
  return self.indexOf(value) === index;
}

function concatTags(tags, store) {
  return [...tags, ...store.tags];
}

const tagsByCategory = createSelector(storesByCategory, stores =>
  stores.reduce(concatTags, []).filter(onlyUnique)
);

/**
 * Add store info (is opened and next schedule day) and sort by is opened
 */
const storesWithOpeningInfoByCategory = createSelector(storesByCategory, stores =>
  stores.map(mapStore).sort(sortByOpened)
);

/**
 * Filter store by Tag
 */
const storesByTag = createSelector(storesWithOpeningInfoByCategory, tagSelector, (stores, tag)=> {
  if (!tag) {
    return stores;
  }

  return stores.filter(({ tags }) => tags.indexOf(tag) !== -1);
});

const makeStoresByCategory = () => storesByTag;
const makeTagsByCategory = () => tagsByCategory;

export const selectors = {
  makeStoresByCategory,
  makeTagsByCategory,
  loading: loadingSelector,
  error: errorSelector,
  tag: tagSelector,
};

export const actions = {
  getStoresOf,
  changeTag
};
