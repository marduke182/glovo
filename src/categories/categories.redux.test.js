import thunk from 'redux-thunk';
import { applyMiddleware, createStore, compose, combineReducers } from 'redux';

import glovoClient from '../shared/libs/glovoClient';
import reducer, { actions, selectors, name } from './categories.redux';
import { actions as storeActions } from '../stores/stores.redux';

jest.mock('../shared/libs/glovoClient');
jest.mock('../stores/stores.redux');


function setup(initialState = {}) {
  const middlewares = [thunk];

  const enhancers = [applyMiddleware(...middlewares)];
  const store = createStore(
    combineReducers({ [name]: reducer }),
    initialState,
    compose(...enhancers)
  );

  return {
    store
  };
}

beforeEach(async () => {
  jest.spyOn(storeActions, 'getStoresOf').mockImplementation(() => async () => {});
});

test('should have initial state', async () => {
  const { store } = setup();
  expect(store.getState()).toMatchSnapshot();
});

describe('get categories', () => {
  beforeEach(async () => {
    glovoClient.get.mockClear();
  });

  test('should show loading', async () => {
    glovoClient.get.mockReturnValueOnce(new Promise(() => {}));
    const { store } = setup();

    store.dispatch(actions.getCategoriesAndStores());

    expect(selectors.loading(store.getState())).toBe(true);
  });

  describe('success', () => {

    const categoriesResponse = {
      data: {
        categories: [
          {
            id: 3,
            label: 'Snacks',
            name: 'snacks',
            openIcon: 'http://localhost:3000/img/snacks-open.png',
            sleepIcon: 'http://localhost:3000/img/snacks-sleep.png'
          },
          {
            id: 2,
            label: 'Gifts',
            name: 'gifts',
            openIcon: 'http://localhost:3000/img/gifts-open.png',
            sleepIcon: 'http://localhost:3000/img/gifts-sleep.png'
          },
          {
            id: 1,
            label: 'Restaurants',
            name: 'restaurants',
            openIcon: 'http://localhost:3000/img/restaurants-open.png',
            sleepIcon: 'http://localhost:3000/img/restaurants-sleep.png'
          },
          {
            id: 4,
            label: 'Wonders',
            name: 'wonders',
            openIcon: 'http://localhost:3000/img/wonders-open.png',
            sleepIcon: 'http://localhost:3000/img/wonders-sleep.png'
          }
        ]
      }
    };

    beforeEach(async () => {
      glovoClient.get.mockImplementation(async () => categoriesResponse);
    });

    test('should store categories in store', async () => {

      const { store } = setup();

      await store.dispatch(actions.getCategoriesAndStores());

      expect(selectors.categories(store.getState())).toEqual(categoriesResponse.data.categories);
    });
  });
});
