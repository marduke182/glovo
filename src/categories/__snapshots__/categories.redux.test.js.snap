// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should have initial state 1`] = `
Object {
  "categories": Object {
    "categories": Array [],
    "error": null,
    "loading": false,
  },
}
`;
